sign1.go

```diff
@@ -250,6 +291,22 @@ func Sign1(rand io.Reader, signer Signer, headers Headers, payload []byte, exter
 	return msg.MarshalCBOR()
 }
 
+// Sign1Detached signs a Sign1Message using the provided Signer.
+//
+// This method is a wrapper of `Sign1Message.SignDetached()`.
+//
+// Reference: https://datatracker.ietf.org/doc/html/rfc8152#section-4.4
+func Sign1Detached(rand io.Reader, signer Signer, headers Headers, detached, external []byte) ([]byte, error) {
+	msg := Sign1Message{
+		Headers: headers,
+	}
+	err := msg.SignDetached(rand, detached, external, signer)
+	if err != nil {
+		return nil, err
+	}
+	return msg.MarshalCBOR()
+}
```

Comment:
Do we really need
```go
func (m *SignMessage) SignDetached(rand io.Reader, detached, external []byte, signers ...Signer) error
func (m *Sign1Message) SignDetached(rand io.Reader, detached, external []byte, signer Signer) error
```
?

Can `Sign1Detached` be implemented as
```suggestion
func Sign1Detached(rand io.Reader, signer Signer, headers Headers, detached, external []byte) ([]byte, error) {
	msg := Sign1Message{
		Headers: headers,
		Payload: detached,
	}
	err := msg.Sign(rand, external, signer)
	if err != nil {
		return nil, err
	}
	msg.Payload = nil
	return msg.MarshalCBOR()
}
```
?

Same question for `SignDetached`.

Reply:
Seems fussy to me, would making detached the default case and calling from `Sign1` be ok? Saves unsetting the payload on the detached object.

Reply:
> would making detached the base case and calling from Sign1 be ok?

IMO yes. @shizhMSFT wdyt?

Reply:
@shizhMSFT, I believe this is the blocking issue on this PR. Can we consider refactoring the implementation, either in this PR, or a separate PR? Either way, can we come to some conclusion here to move forward?

Reply:
@alex-richards Are you suggesting
```go
func sign1(rand io.Reader, signer Signer, headers Headers, payload, external []byte) (*Sign1Message, error) {
	msg := Sign1Message{
		Headers: headers,
		Payload: payload,
	}
	err := msg.Sign(rand, external, signer)
	if err != nil {
		return nil, err
	}
	return msg, nil
}

func Sign1(rand io.Reader, signer Signer, headers Headers, payload, external []byte) ([]byte, error) {
	msg, err := sign1(rand, signer, headers, payload, external)
	if err != nil {
		return nil, err
	}
	return msg.MarshalCBOR()
}

func Sign1Detached(rand io.Reader, signer Signer, headers Headers, detached, external []byte) ([]byte, error) {
	msg, err := sign1(rand, signer, headers, detached, external)
	if err != nil {
		return nil, err
	}
	msg.Payload = nil
	return msg.MarshalCBOR()
}
```
That refactoring works for me.

Reply:
@alex-richards, @shizhMSFT 
Can this comment be resoled?

Reply:
@SteveLasker I think it will be better if we gather more inputs from other maintainers.

Reply:
Disclaimer: I am very much a random passerby, nudged to look at this. I do spend a lot time with COSE, in languages other than Go.

The refactoring proposed by @shizhMSFT seems quite reasonable to me, and results in a slightly cleaner API in my opinion.

Reply:
Ping on this change for closure...